test/boost/aws_error_injection_test.cc

```diff
@@ -0,0 +1,195 @@
+        }
+
+    private:
+        sstring build_response(const http::request& req) {
+            sstring response;
+            if (req._method == "POST") {
+                if (req.query_parameters.contains("uploads")) {
+                    response = "<InitiateMultipartUploadResult><Bucket>bucket</Bucket><Key>key</Key><UploadId>UploadId</UploadId></"
+                               "InitiateMultipartUploadResult>";
+                } else if (req.query_parameters.contains("uploadId")) {
+                    switch (test_failure_policy) {
+                    case failure_policy::SUCCESS:
+                        return R"(<CompleteMultipartUploadResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
+                                     <Location>http://Example-Bucket.s3.Region.amazonaws.com/Example-Object</Location>
+                                     <Bucket>Example-Bucket</Bucket>
+                                     <Key>Example-Object</Key>
+                                     <ETag>"3858f62230ac3c915f300c664312c11f-9"</ETag>
+                                </CompleteMultipartUploadResult>)";
+                    case failure_policy::RETRYABLE_FAILURE:
+                        // should succeed on retry
+                        test_failure_policy = failure_policy::SUCCESS;
+                        return R"(<?xml version="1.0" encoding="UTF-8"?>
+
+                                 <Error>
+                                  <Code>InternalError</Code>
+                                  <Message>We encountered an internal error. Please try again.</Message>
+                                  <RequestId>656c76696e6727732072657175657374</RequestId>
+                                  <HostId>Uuag1LuByRx9e6j5Onimru9pO4ZVKnJ2Qz7/C1NPcfTWAtRPfTaOFg==</HostId>
+                                 </Error>)";
+                    case failure_policy::NONRETRYABLE_FAILURE:
+                        return R"(<?xml version="1.0" encoding="UTF-8"?>
+
+                                 <Error>
+                                  <Code>InvalidAction</Code>
+                                  <Message>We encountered an internal error. Please try again.</Message>
+                                  <RequestId>656c76696e6727732072657175657374</RequestId>
+                                  <HostId>Uuag1LuByRx9e6j5Onimru9pO4ZVKnJ2Qz7/C1NPcfTWAtRPfTaOFg==</HostId>
+                                 </Error>)";
+                    }
+                }
+            }
+            return response;
+        }
+        failure_policy test_failure_policy;
+    };
+    explicit server(failure_policy _test_failure_policy) : test_failure_policy(_test_failure_policy) {}
+    future<> start() {
+        co_await seastar::async([this] {
+            uint16_t port = 2012;
+            net::inet_address addr("127.0.0.1");
+            test_server.start("test").get();
+
+            test_server.server()
+                .invoke_on_all([this](httpd::http_server& server) {
```

Comment:
I believe that proxy server with fine-grained of what request fails and what is (!) _forwarded_ to existing minio instance (or whatever is configured as S3 server via environment) would be needed anyway

Reply:
Agree, but it requires more investigation to create real life cases because not every S3 operation can return any of these errors, also it makes sense to unify response body and http errors handling

Reply:
A suggestion -- make _existing_ client with _existing_ minio server just retry _any_ request two times regardless of what (with a on/off switch of course). Since we mostly care about the case when server served the request, but failed to deliver back the response (or may client failed to receive the delivered response, whatever), this should be the exact emulation we're looking for

Reply:
https://github.com/scylladb/scylladb/issues/20613